src/libraries/System.Runtime.Numerics/src/System/Numerics/BigIntegerCalculator.DivRem.cs

```diff
@@ -217,72 +179,65 @@ public static unsafe uint[] Remainder(uint[] left, uint[] right)
                 if (digit > 0)
                 {
                     // Now it's time to subtract our current quotient
-                    uint carry = SubtractDivisor(left + n, leftLength - n,
-                                                 right, rightLength, digit);
+                    uint carry = SubtractDivisor(left.Slice(n), right, digit);
                     if (carry != t)
                     {
                         Debug.Assert(carry == t + 1);
 
                         // Our guess was still exactly one too high
-                        carry = AddDivisor(left + n, leftLength - n,
-                                           right, rightLength);
+                        carry = AddDivisor(left.Slice(n), right);
                         --digit;
 
                         Debug.Assert(carry == 1);
                     }
                 }
 
                 // We have the digit!
-                if (bitsLength != 0)
-                    bits[n] = (uint)digit;
-                if (i < leftLength)
-                    left[i] = 0;
+                if (!bits.IsEmpty)
+                    Unsafe.Add(ref resultPtr, n) = (uint)digit;
```

Comment:
There are some places when JIT is not smart enough. Let's take a look at BigIntegerCalculator.AddSub.cs file:
```csharp
            for ( ; i < right.Length; i++)
            {
                long digit = (Unsafe.Add(ref leftPtr, i) + carry) + right[i];
                Unsafe.Add(ref resultPtr, i) = unchecked((uint)digit);
                carry = digit >> 32;
            }
            for ( ; i < left.Length; i++)
            {
                long digit = left[i] + carry;
                Unsafe.Add(ref resultPtr, i) = unchecked((uint)digit);
                carry = digit >> 32;
            }
```
I don't see the way how JIT is able to eliminate bounds check for `left` in the first loop. The same for the second loop.

Reply:
Right. There are some places where the JIT still cannot elide it, such as in your examples where multiple spans would be accessed within a single loop construct. And I wouldn't change those from what you have. But in the specific case I called out here, which involves only a single span access within the if block, the JIT can elide the bounds check just fine in a fully safe mechanism.

Reply:
Here's an alternative approach which can elide bounds checks if needed.

```cs
using System;

public class C {
    public static int Sum1(int i, Span<int> span) {
        int sum = 0;
        for (; i < span.Length; i++)
        {
            sum += span[i];
        }
        return sum;
    }
    
    public static int Sum2(int i, Span<int> span) {
        int sum = 0;
        foreach (int value in span.Slice(i))
        {
            sum += value;
        }
        return sum;
    }
    
    public static void Add1(int i, Span<int> span) {
        for (; i < span.Length; i++)
        {
            span[i] += 1;
        }
    }
    
    public static void Add2(int i, Span<int> span) {
        foreach (ref int value in span.Slice(i))
        {
            value += 1;
        }
    }
}
```

In both _Sum1_ and _Add1_, the JIT emits a bounds check inside every iteration of the loop. The reason for this is that the JIT primarily understands the pattern `for (int i = 0; i < span.Length; i++)`, and if this pattern is modified (e.g., _i_ does not start with a known constant non-negative value), the JIT won't optimize it.

However, inside _Sum2_ and _Add2_, the bounds check is emitted only once, during the call to `Slice`. All iterations within the loop are completely free of bounds checks, as a _foreach_ iterator over a span (including a sliced span!) also matches a recognized JIT pattern. If within the iterator block you capture the iterator variable as a _ref T_ instead of a _T_, you'll also be able to write back to it within that block.

Depending on how the loops are set up, this could be a mechanism to get fully-safe code with equivalent performance to unsafe code, modulo the single bounds check at the very beginning.

Reply:
The last commit 020e749 reduces the number of unsafe code when it can be safely replaced with indexer.